group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java

```diff
@@ -426,9 +429,43 @@ public CompletableFuture<ListGroupsResponseData> listGroups(
             return FutureUtils.failedFuture(Errors.COORDINATOR_NOT_AVAILABLE.exception());
         }
 
-        return FutureUtils.failedFuture(Errors.UNSUPPORTED_VERSION.exception(
-            "This API is not implemented yet."
-        ));
+        List<CompletableFuture<ListGroupsResponseData>> futures = new java.util.ArrayList<>(Collections.emptyList());
+        for (int i = 0; i < numPartitions; i++) {
+            futures.add(runtime.scheduleReadOperation("list_groups",
+                    new TopicPartition(Topic.GROUP_METADATA_TOPIC_NAME, i),
+                    (coordinator, __) -> coordinator.listGroups(context, request)
```

Comment:
We need to use the second parameters `__` as pass it down to `listGroups`. For the context, the second parameter is the last committed offsets. We should list the groups based on it. Otherwise, we would return uncommitted changes.

Reply:
I don’t know if I understand it right, here we need to replace __ with lastCommittedOffset, right?

Reply:
Right. Then you must pass `lastCommittedOffset` to `listGroups` as well and use it to query the timeline data structures.

Reply:
Basically [here](https://github.com/apache/kafka/pull/14271/files#diff-00f0f81cf13e66781777d94f7d2e68a581663385c37e98792507f2294c91bb09R418), you need to use `lastCommittedOffset` to ensure that you only read committed state.